### DISCUSSION TITLE
displaying values for points on mouse hover

### DISCUSSION THREAD

**Discussion by jkissi** [marcomusy/vedo#552] [Q&A] [answered]
Hi!! 
I would like to display values for the points in my point cloud when the mouse hovers over them. I was wondering if you had any examples that specifically show how to do that?

cheers
j

**Comment by marcomusy** [comment, marked as answer]
Hi
yes - check out these 2 examples:
`vedo -r hoverLegend`
`vedo -r mousehover`

You can search example by typing e.g. "hover" in the search field at
https://vedo.embl.es/#gallery

**Reply by jkissi** [reply]
Thanks, these are very helpful examples. I have general question about callbacks:
is it possible to pass specific arguments to "func" in the  .addCallback() function? I would like to separate the callback logic from the plotting logic in order to keep things a little bit tidyer. For example, in mousehover.py example:

```
    def func(evt):                       ### called every time mouse moves!
        msh = evt.actor
        if not msh: return               # mouse hits nothing, return.
        pt = evt.picked3d                # 3d coords of point under mouse

        pid = msh.closestPoint(pt, returnPointId=True)
        txt = f"Point:  {precision(pt[:2]  ,2)}\n" \
              f"Height: {precision(arr[pid],3)}\n" \
              f"Ground speed: {precision(evt.speed3d*100,2)}"
        arw = Arrow(pt - evt.delta3d, pt, s=0.001, c='orange5')
        vig = msh.vignette(txt, point=pt, offset=(0.4,0.6),
                           s=0.04, c='k', font="VictorMono").followCamera()
        msg.text(txt)                    # update text message
        if len(plt.actors)>3: plt.pop()  # remove the old vignette
        plt.add([arw, vig])              # add Arrow and the new vignette

    msg = Text2D(pos='bottom-left', font="VictorMono") # an empty text
    hil = ParametricShape('RandomHills').cmap('terrain').addScalarBar()
    arr = hil.pointdata["Scalars"]       # numpy array with heights

    plt = Plotter(axes=1, bg2='lightblue')
    plt.addCallback('mouse move', func)  # add the callback function
    plt.addCallback('keyboard', lambda e: plt.remove(plt.actors[3:]).render())
    plt.show(hil, msg, __doc__, viewup='z')
```
I would like to move func() to a separate file for but to do that [I think] I have to be able to pass multiple objects into it. Can I do this? is it necessary? Or does the callback function always have to be in the same scope as the plot object?

Cheers
j

**Reply by marcomusy** [reply]
I guess you can place the `func` in a different file and then import it in the main (along with variables that `func` needs). E.g.
In mymodule.py:
```python
var1 = 1
var2 = 2

def func(evt):
    print(var2)
```

In main.py
```python
import mymodule
...
mymodule.var2= 3
...
plt.addCallback('mouse move', mymodule.func)  # add the callback function
```

but you cannot pass arguments in the `addCallback´.

**Reply by jkissi** [reply]
OK, I understand. As always, very, very helpful, thank you!!!
j
